### Import Library and dataset

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [2]:
df = pd.read_csv('dataset.csv')
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  sales  \
0                   3              0     1                      0  sales   
1                   6              0     1                      0  sales   
2                   4              0     1                      0  sales   
3                   5              0     1                      0  sales   
4                   3              0     1                      0  sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low


<br>


In [3]:
df.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'sales', 'salary'],
      dtype='object')

### Data Preprocessing

##### Data Info

In [3]:
df.shape

(14999, 10)

This dataset contains 10 columns (features) and 14999 rows (records)

##### Column list

In [4]:
df.columns.to_list()

['satisfaction_level',
 'last_evaluation',
 'number_project',
 'average_montly_hours',
 'time_spend_company',
 'Work_accident',
 'left',
 'promotion_last_5years',
 'sales',
 'salary']

##### dataframe info

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


Here sales and salary columns are object (string) datatype

##### Missing Value Handing - Checking Null Value

In [6]:
df.isna().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
sales                    0
salary                   0
dtype: int64

There is no null value (NaN)

##### Data statistics

In [7]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

<br>

##### Chnage the column name

In [8]:
df = df.rename(columns={'sales' : 'department'})

In [9]:
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

##### Change the columns values 

The 'department' column of the dataset has many categories, need to reduce the categories for better modelling. 
Let’s see all the categories of the department column:

In [10]:
df['department'].unique()

array(['sales', 'accounting', 'hr', 'technical', 'support', 'management',
       'IT', 'product_mng', 'marketing', 'RandD'], dtype=object)

Here we can merge “technical”, “support” columns into one column to make our analysis easier.

In [11]:
df['department']=np.where(df['department'] =='support', 'technical', df['department'])

Changed column value of “support” and “IT” into "Technical"

<br>

### Categorical to Numerical  Transformation

##### Dummy Variable Encoding

In [12]:
depart = pd.get_dummies(df['department'], prefix='department', drop_first=True )
sales = pd.get_dummies(df['salary'], prefix='salary', drop_first=True )

In [13]:
df = df.join(depart)
df = df.join(sales)

Now the actual variables need to be removed after the dummy variable have been created. Column names after creating dummy variables for categorical variables:

In [14]:
cols = ['department', 'salary']
df.drop(cols, axis=1, inplace=True)

In [15]:
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  \
0                   3              0     1                      0   
1                   6              0     1                      0   
2                   4              0     1                      0   
3                   5              0     1                      0   
4                   3              0     1                      0   

   department_RandD  department_accounting  department_hr  \
0                 0                      0              0   
1                 0                      0              0   
2                 0                      0              0   
3                 0                      0              0   
4                 0                      0              0   

   department_management  department_marketing  department_product_mng  \
0                      0                     0                       0   
1                      0                     0                       0   
2                      0                     0                       0   
3                      0                     0                       0   
4                      0                     0                       0   

   department_sales  department_technical  salary_low  salary_medium  
0                 1                     0           1              0  
1                 1                     0           0              1  
2                 1                     0           0              1  
3                 1                     0           1              0  
4                 1                     0           1              0

In [16]:
df.shape

(14999, 18)

Now this dataset contains 10 columns (features) and 14999 rows (records)

<br>

### Dependent and Independent Variables

In [17]:
X = df.drop('left', axis=1)       # X is complete dataframe except "left" column 
y = df['left']                    # y holds the "left" column

In [19]:
len(X.columns)

17

In [22]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: left, dtype: int64

In [24]:
X.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident',
       'promotion_last_5years', 'department_RandD', 'department_accounting',
       'department_hr', 'department_management', 'department_marketing',
       'department_product_mng', 'department_sales', 'department_technical',
       'salary_low', 'salary_medium'],
      dtype='object')

##### selecting the required columns form the 17 column

In [25]:
new_cols = ['satisfaction_level', 'last_evaluation', 'time_spend_company', 'Work_accident', 'promotion_last_5years', 'department_RandD',
       'department_hr', 'department_management', 'salary_low', 'salary_medium']

In [27]:
X = df[new_cols]
y = df['left']

<br>

### Train and Test Split

In [28]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state=92)

### Model Building

#### Logistics Regression

In [29]:
from sklearn.linear_model import LogisticRegression

model_logistic = LogisticRegression()

In [31]:
model_logistic.fit(X_train, y_train)

LogisticRegression()

In [32]:
model_logistic.score(X_train, y_train)

0.7702638346509192

In [33]:
model_logistic.score(X_test, y_test)

0.7664444444444445

The accuracy score of the logistic regression is about 0.76 , which is not so good.

<br>

#### Decision Tree

In [34]:
from sklearn.tree import DecisionTreeClassifier
model_tree = DecisionTreeClassifier()

In [35]:
model_tree.fit(X_train, y_train)

DecisionTreeClassifier()

In [36]:
model_tree.score(X_train, y_train)

0.9967615963425088

In [37]:
model_tree.score(X_test, y_test)

0.9702222222222222

The accuracy score of the Decision Tree model is about 0.97. 

<br>

#### Random Forest 

In [38]:
from sklearn.ensemble import RandomForestClassifier

model_rf = RandomForestClassifier()

In [39]:
model_rf.fit(X_train, y_train)

RandomForestClassifier()

In [40]:
model_rf.score(X_train, y_train)

0.9965711020097152

In [41]:
model_rf.score(X_test, y_test)

0.9864444444444445

The accuracy score of the Random Forest model is about 0.98. 

<br>

So, as we can see that the Random Forest Model has proven to be more useful in the prediction of employee turnover, now let’s have a look at the feature importance of our random forest classification model.

In [42]:
model_rf.feature_importances_

array([0.50611487, 0.18339367, 0.26617228, 0.01495086, 0.00279437,
       0.00291578, 0.00329404, 0.00304832, 0.01200081, 0.00531499])

In [43]:
feature_labels = np.array(['satisfaction_level', 'last_evaluation', 'time_spend_company', 'Work_accident', 'promotion_last_5years', 
      'department_RandD', 'department_hr', 'department_management', 'salary_high', 'salary_low'])
importance = model_rf.feature_importances_
feature_indexes_by_importance = importance.argsort()
for index in feature_indexes_by_importance:
    print('{}    -> {:.2f}%'.format(feature_labels[index], (importance[index] *100.0)))

promotion_last_5years    -> 0.28%
department_RandD    -> 0.29%
department_management    -> 0.30%
department_hr    -> 0.33%
salary_low    -> 0.53%
salary_high    -> 1.20%
Work_accident    -> 1.50%
last_evaluation    -> 18.34%
time_spend_company    -> 26.62%
satisfaction_level    -> 50.61%


According to our Random Forest classification model, the above aspects show the most important features which will influence whether an employee will leave the company, in ascending order.

In [1]:

import numpy as np
impo=[0.50611487, 0.18339367, 0.26617228, 0.01495086, 0.00279437,0.00291578, 0.00329404, 0.00304832, 0.01200081, 0.00531499]
feature_labels = ['satisfaction_level', 'last_evaluation', 'time_spend_company', 'Work_accident', 'promotion_last_5years', 
      'department_RandD', 'department_hr', 'department_management', 'salary_high', 'salary_low']


In [2]:
feature_labels

array(['satisfaction_level', 'last_evaluation', 'time_spend_company',
       'Work_accident', 'promotion_last_5years', 'department_RandD',
       'department_hr', 'department_management', 'salary_high',
       'salary_low'], dtype='<U21')

In [1]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9
